add support for ltc4162-f/s and ltc4015 battery charger

kseerp opened the pull request:
## PR Description
Add support for:
LTC4162-F 35V/3.2A Multi-Cell LiFePO4 Step-Down Battery Charger
LTC4162-S 35V/3.2A Lead-Acid Step-Down Battery Charger
LTC4015 35V/3.2A Multichemistry Buck Battery Charger Controller

Bindings:
- Add compatible entries for ltc4162-f/s and ltc4015
- Include datasheets for new devices
- Change vendor prefix from lltc to adi

ltc4162:
- Modify functions for battery voltage/current, input voltage/current,
charge voltage, die temp, and force telemetry to handle different
battery chemistries.

## PR Type
- [x] New feature (a change that adds new functionality)
- [ ] Bug fix (a change that fixes an issue)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

nunojsa commented on Documentation/devicetree/bindings/power/supply/ltc4162-l.yaml:
```diff
@@ -47,8 +53,8 @@ properties:
-  - lltc,rsnsb-micro-ohms
-  - lltc,rsnsi-micro-ohms
+  - adi,rsnsb-micro-ohms
+  - adi,rsnsi-micro-ohms
```
I do not think that changing the prefix will be accepted by the maintainers. That will effectively break backward compatibility...

nunojsa commented on drivers/power/supply/ltc4162-l-charger.c:
```diff
@@ -108,6 +120,19 @@ static u8 ltc4162l_get_cell_count(struct ltc4162l_info *info)
+		return ret;
+
+	/* Upper 4 bits is the chemistry type */
+	return val >> 8 & 0x0f;
```
`FIELD_GET()`... If that's not the style using in the driver, a precursor patch could be a nice to have

nunojsa commented on drivers/power/supply/ltc4162-l-charger.c:
```diff
@@ -589,6 +738,9 @@ static ssize_t force_telemetry_show(struct device *dev,
 		return ret;
 
+	if (info->type == ltc4015)
+		return sprintf(buf, "%u\n", regval & BIT(4) ? 1 : 0);
```
another nice cleanup patch would be to use `sysfs_emit()`

nunojsa commented on drivers/power/supply/ltc4162-l-charger.c:
```diff
@@ -846,33 +1018,40 @@ static int ltc4162l_probe(struct i2c_client *client,
 	}
 
-	ret = device_property_read_u32(dev, "lltc,rsnsb-micro-ohms",
+	type = (enum ltc4162l_type)(uintptr_t)device_get_match_data(dev);
```
Hmm, please don't. In this case you can't even check for errors... Please use a dedicated `chip_info` structure. That will help with other things... see below.

nunojsa commented on drivers/power/supply/ltc4162-l-charger.c:
```diff
@@ -846,33 +1018,40 @@ static int ltc4162l_probe(struct i2c_client *client,
+							   &ltc4162l_config);
+	else
+		info->charger = devm_power_supply_register(dev, &ltc4162l_desc,
+							   &ltc4162l_config);
```
Just have the proper _desc in the chip_info struct and pass that directly in here. No need for this `if()`

nunojsa commented on drivers/power/supply/ltc4162-l-charger.c:
```diff
@@ -920,12 +1105,12 @@ static struct i2c_driver ltc4162l_driver = {
 	.id_table	= ltc4162l_i2c_id_table,
 	.driver = {
-		.name		= "ltc4162-l-charger",
+		.name		= "ltc4162-charger",
```
Not sure but this could also be an ABI breakage. To be on the safe side, don't change the driver name...

nunojsa commented:
Also, the things I mentioned as "nice to have" is up to you to add it or not :)

kseerp commented:
v2
- Added chip_info struct to hold chip specific data.
- Added functions specific to each chip to get/set vbat, vcharge, die_temp.
- Duplicated the default descriptor to configure name based on chip_info.
- Used FIELD_GET to extract the value from register.
- Changed sprintf to sysfs_emit.

nunojsa commented on drivers/power/supply/ltc4162-l-charger.c:
```diff
@@ -840,6 +1153,12 @@ static int ltc4162l_probe(struct i2c_client *client,
 	info->client = client;
 	i2c_set_clientdata(client, info);
 
+	chip_info = device_get_match_data(dev);
```
when upstreaming use [i2c_get_match_data()](https://elixir.bootlin.com/linux/v6.12-rc3/C/ident/i2c_get_match_data)

nunojsa commented on drivers/power/supply/ltc4162-l-charger.c:
```diff
@@ -820,12 +1131,14 @@ static void ltc4162l_clear_interrupts(struct ltc4162l_info *info)
 
 static int ltc4162l_probe(struct i2c_client *client,
-			const struct i2c_device_id *id)
+			  const struct i2c_device_id *id)
```
This is a stylistic change... You have more in this patch. Do them in a different patch (like a follow up one)

nunojsa commented on drivers/power/supply/ltc4162-l-charger.c:
```diff
@@ -628,14 +897,14 @@ static ssize_t arm_ship_mode_show(struct device *dev,
-	size_t count)
+				   struct device_attribute *attr,
+				   const char *buf,
+				   size_t count)
```
style...

nunojsa commented on drivers/power/supply/ltc4162-l-charger.c:
```diff
@@ -628,14 +897,14 @@ static ssize_t arm_ship_mode_show(struct device *dev,
-	return sprintf(buf, "%u\n",
-		regval == LTC4162L_ARM_SHIP_MODE_MAGIC ? 1 : 0);
+	return sysfs_emit(buf, "%u\n",
+			  regval == LTC4162L_ARM_SHIP_MODE_MAGIC ? 1 : 0);
```
also not related

nunojsa commented on drivers/power/supply/ltc4162-l-charger.c:
```diff
@@ -607,17 +877,16 @@ static ssize_t force_telemetry_store(struct device *dev,
+				 value ? info->chip_info->telemetry_mask : 0);
 
-	return count;
+	return ret ? ret : count;
```
ditto

nunojsa commented on drivers/power/supply/ltc4162-l-charger.c:
```diff
@@ -98,7 +138,7 @@ static u8 ltc4162l_get_cell_count(struct ltc4162l_info *info)
 
 	/* Lower 4 bits is the cell count, or 0 if the chip doesn't know yet */
-	val &= 0x0f;
+	val = FIELD_GET(LTC4162L_CELL_COUNT_MASK, val);
```
This is also not related with addition of the new devices...

nunojsa reviewed (commented):
Code looks mostly good to me. Just make sure to split your changes a bit better so it makes it easier to review. At this point, I would say to send this upstream...

kseerp commented:
Series applied upstream. Add ADI compliant version
